proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1012,6 +1012,268 @@ lemma refillUpdate_corres:
    apply clarsimp+
   done
 
+lemma head_insufficient_length_cross:
+  "\<lbrakk>(s,s') \<in> state_relation; valid_objs s; pspace_aligned s; pspace_distinct s;
+    pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg))
+             (sc_refill_cfgs_of s) scPtr;
+    the (head_insufficient scPtr s); valid_objs' s'\<rbrakk>
+   \<Longrightarrow> obj_at' (\<lambda>sc'. Suc 0 < scRefillCount sc') scPtr s'"
+  apply (frule (1) head_insufficient_length_at_least_two[rotated])
```

Comment:
Hmm. This is a rather odd crossing lemma. Crossing the information in the assumption would give a lot more than the conclusion here. I think this should be two separate lemmas. One for proving that `head_insufficient` implies "refill length greater than 1" on the abstract side, which is actually proved below. Then, the second one is a lemma for crossing the conditions on the refill length, which would require a bit more than simply unfolding the `sc_relation`. I am not sure if we already have such a lemma.

Reply:
Actually, I did write this length crossing lemma myself when I proved `refillUnblockCheck_corres`..
```
lemma length_sc_refills_cross:
  "\<lbrakk>(s, s') \<in> state_relation; sc_at scp s; sc_at' scp s'; valid_refills'' scp s';
   ((\<lambda>sc. P (length (sc_refills sc))) |< scs_of2 s) scp\<rbrakk>
   \<Longrightarrow> ((\<lambda>sc'. P (scRefillCount sc')) |< scs_of' s') scp"
  apply (drule (2) state_relation_sc_relation)
  apply (clarsimp simp: obj_at_simps is_sc_obj valid_refills''_def sc_relation_def opt_map_left_Some)
  done
  ```
  this is in https://github.com/mktnk3/l4v/commit/d92215ab52ff919fa77149bd2490f9a25b86e9df

Reply:
Oh cool. I'll use that

Reply:
This can be made into an if and only if, which has been helpful. I'll include this vesion of this rule, if you're OK with that